### PR TITLE
default filters as routing.PreProcessors

### PR DESCRIPTION
**Pull request by szuecs**
Implements #1002
Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

### CODE REVIEW

File: cmd/skipper/defaultfilterflags.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+type defaultFiltersFlags struct {
+	filters []*eskip.Filter
+}
+
+func (dpf *defaultFiltersFlags) String() string {
+	res := make([]string, len(dpf.filters))
+	for _, f := range dpf.filters {
+		res = append(res, f.Name)
```
Comment (aryszka) [inline cmd/skipper/defaultfilterflags.go]: this will print sg like: `foo -> bar -> baz`. Where is it used?

File: skipper.go [hunk 2]
```diff
@@ -861,6 +865,7 @@ func Run(o Options) error {
 			loadbalancer.HealthcheckPostProcessor{LB: lbInstance},
 			loadbalancer.NewAlgorithmProvider(),
 		},
+		PreProcessors:   []routing.PreProcessor{o.DefaultFilters},
```
Comment (aryszka) [inline skipper.go]: we should not add the o.DefaultFilters when it is nil

File: cmd/skipper/defaultfilterflags.go (hunk 1 above)
Comment (szuecs) [reply cmd/skipper/defaultfilterflags.go]: only if you print the flags, so I don't really care

File: eskip/eskip.go [hunk 3]
```diff
@@ -20,6 +20,20 @@ var (
 	duplicateMethodPredicateError   = errors.New("duplicate method predicate")
 )
 
+type DefaultFilters struct {
+	Prepend []*Filter
+	Append  []*Filter
+}
+
+func (df *DefaultFilters) Do(r []*Route) []*Route {
+	for i := range r {
```
Comment (aryszka) [inline eskip/eskip.go]: we should clone the routes objects. The reason is that a DataClient may cache parsed routes, and this logic changes the Filters field on those routes, too. This way we may end up prepending and appending to the same slice of filters.

File: skipper.go (hunk 2 above)
Comment (szuecs) [reply skipper.go]: fixed

File: routing/datasource.go [hunk 4]
```diff
@@ -494,6 +494,11 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
 		select {
 		case defs := <-updatesRelay:
 			o.Log.Info("route settings received")
+
+			for i := range o.PreProcessors {
```
Comment (aryszka) [inline routing/datasource.go]: it would be nice to have a test checkes that they are applied

File: eskip/eskip.go [hunk 5]
```diff
@@ -20,6 +20,20 @@ var (
 	duplicateMethodPredicateError   = errors.New("duplicate method predicate")
 )
 
+type DefaultFilters struct {
+	Prepend []*Filter
+	Append  []*Filter
+}
+
+func (df *DefaultFilters) Do(r []*Route) []*Route {
```
Comment (aryszka) [inline eskip/eskip.go]: it would be nice to have a test that the filters are prepended and appended

File: eskip/eskip.go [hunk 6]
```diff
@@ -20,6 +20,20 @@ var (
 	duplicateMethodPredicateError   = errors.New("duplicate method predicate")
 )
 
+type DefaultFilters struct {
```
Comment (aryszka) [inline eskip/eskip.go]: A short godoc could explain that this is a PreProcessor and to be used with routing package.

File: eskip/eskip.go (hunk 3 above)
Comment (szuecs) [reply eskip/eskip.go]: done

File: eskip/eskip.go (hunk 5 above)
Comment (szuecs) [reply eskip/eskip.go]: done

File: cmd/skipper/defaultfilterflags.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	defaultPrependFiltersUsage = "set of default filters to apply to prepend to all filters of all routes"
+	defaultAppendFiltersUsage  = "set of default filters to apply to append to all filters of all routes"
+)
+
+type defaultFiltersFlags struct {
+	filters []*eskip.Filter
+}
+
+func (dpf *defaultFiltersFlags) String() string {
+	res := make([]string, len(dpf.filters))
```
Comment (lmineiro) [inline cmd/skipper/defaultfilterflags.go]: shouldn't this be `make([]string, 0, len(dpf.filters))`?

Comment (lmineiro) [reply cmd/skipper/defaultfilterflags.go]: The append below will start appending elements after `len(dpf.filters)`

Comment (aryszka) [reply cmd/skipper/defaultfilterflags.go]: :+1:

File: routing/datasource.go (hunk 4 above)
Comment (szuecs) [reply routing/datasource.go]: done

File: cmd/skipper/defaultfilterflags.go (hunk 7 above)
Comment (szuecs) [reply cmd/skipper/defaultfilterflags.go]: it's not different if you have no `if ... continue` or `break`

Comment (aryszka) [reply cmd/skipper/defaultfilterflags.go]: it should be also fine if we just remove this method

Comment (aryszka) [reply cmd/skipper/defaultfilterflags.go]: my bad, must implement Stringer

Comment (szuecs) [reply cmd/skipper/defaultfilterflags.go]: we decided to drop this and just return a static string instead

Comment (aryszka) [reply cmd/skipper/defaultfilterflags.go]: @szuecs if it was only about the initialization, you'd be right. The comment is about how the append() will behave afterwards

Comment (szuecs) [reply cmd/skipper/defaultfilterflags.go]: https://play.golang.org/p/GSEH2W1uANR @lmineiro you are right

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Pull request merged**